File: 백준/Gold/17141. 연구소 2/연구소 2.java
import java.io.*;
import java.util.*;

public class Main {

  static int N, M, empty, ans = 2500;
  static int[][] lab;
  static List<Point> virus = new ArrayList<>();

  public static void main(String[] args) throws Exception {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());
    N = readInt(st);
    M = readInt(st);
    lab = new int[N][N];
    for (int i = 0; i < N; i++) {
      st = new StringTokenizer(br.readLine());
      for (int j = 0; j < N; j++) {
        int v = readInt(st);
        lab[i][j] = v;
        if (v == 1) continue;
        if (v == 2) {
          virus.add(new Point(i, j));
        }
        ++empty;
      }
    }
    combination(-1, 0, new ArrayList<>());
    System.out.println(ans != 2500 ? ans : -1);
  }

  static int bfs(List<Point> arr, int remain) {
    int[] dy = {1, 0, -1, 0};
    int[] dx = {0, 1, 0, -1};
    Deque<Point> q = new ArrayDeque<>();
    remain -= M;
    if (remain == 0) {
      return 0;
    }
    int[][] visited = new int[N][N];
    for (Point p : arr) {
      q.add(p);
      visited[p.y][p.x] = 1;
    }
    int time = 0;
    while (!q.isEmpty()) {
      Point cur = q.pollFirst();
      for (int d = 0; d < 4; d++) {
        int ny = cur.y + dy[d], nx = cur.x + dx[d];
        if (0 <= ny && ny < N && 0 <= nx && nx < N && visited[ny][nx] == 0 && lab[ny][nx] != 1) {
          visited[ny][nx] = visited[cur.y][cur.x] + 1;
          --remain;
          if (visited[ny][nx] > time) {
            time = visited[ny][nx];
          }
          q.add(new Point(ny, nx));
        }
      }
    }
    return remain == 0 ? time - 1 : -1;
  }

  static void combination(int cur, int dep, List<Point> arr) {
    if (dep == M) {
      int time = bfs(arr, empty);
      if (time != -1 && time < ans) {
        ans = time;
      }
      return;
    }
    for (int i = cur + 1; i < virus.size(); i++) {
      arr.add(virus.get(i));
      combination(i, dep + 1, arr);
      arr.remove(dep);
    }
  }

  static int readInt(StringTokenizer st) {
    return Integer.parseInt(st.nextToken());
  }

  static class Point {
    int y, x;

    Point(int y, int x) {
      this.y = y;
      this.x = x;
    }
  }

}

File: 백준/Gold/17141. 연구소 2/연구소 2.py
from collections import deque
import sys
input = sys.stdin.readline
        
def bfs(choose_virus, empty):
    dy = [1, 0, -1, 0]
    dx = [0, 1, 0, -1]
    empty -= len(choose_virus)
    q = deque(choose_virus)
    visited = [[-1] * N for _ in range(N)]
    for r, c in choose_virus:
        visited[r][c] = 0
    time = 0
    while q:
        cr, cc = q.popleft()
        for d in range(4):
            nr, nc = cr + dy[d], cc + dx[d]
            if 0 <= nr < N and 0 <= nc < N and visited[nr][nc] == -1 and lab[nr][nc] != 1:
                visited[nr][nc] = visited[cr][cc] + 1
                empty -= 1
                if visited[nr][nc] > time:
                    time = visited[nr][nc]
                q.append((nr, nc))
    return time if empty == 0 else -1

def combination(cur, dep, max_cnt, arr):
    global ans
    if dep == max_cnt:
        time = bfs(arr, empty_cnt)
        if time != -1 and time < ans:
            ans = time
        return
    for i in range(cur + 1, len(virus)):
        arr.append(virus[i])
        combination(i, dep + 1, max_cnt, arr)
        arr.pop()

N, M = map(int, input().split())
lab = [[*map(int, input().split())] for _ in range(N)]
virus = []
empty_cnt = 0
ans = 1e9
for i in range(N):
    for j in range(N):
        if lab[i][j] == 1:
            continue
        if lab[i][j] == 2:
            virus.append((i, j))
        empty_cnt += 1
combination(-1, 0, M, [])
print(ans if ans != 1e9 else -1)